Olá, 

Meu nome é Ramon. Ao ao longo do texto farei algumas observações sobre melhorias no código. Nas partes em que não houver comentários, está tudo certo, ok? Estarei aberto a feedbacks e discussões sobre o tema.

**Peço que mantenha e não altere os comentários que eu fizer por aqui para que possamos nos localizar posteriormente, ok?**

Mais uma coisa, vamos utilizar um código de cores para você entender os meus feedbacks no seu notebook. Funciona assim:

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Sucesso. Tudo foi feito corretamente.
</div>

<div class="alert alert-block alert-warning">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Alerta não crítco, mas que pode ser corrigido para melhoria geral no seu código/análise
</div>

<div class="alert alert-block alert-danger">

<b>Comentário do revisor: </b> <a class="tocSkip"></a>
    
Erro que precisa ser consertado, caso contrário seu projeto **não** será aceito.
</div>

Você pode me responder usando isso:
<div class="alert alert-block alert-info">
<b>Resposta do Aluno.</b> <a class="tocSkip"></a>
</div>

###




<div class="alert alert-block alert-success">
<b> Comentário Geral do Revisor Final: </b> <a class="tocSkip"></a>

Tudo certo agora. Desejo sucesso na jornada. Parabéns pelo empenho.
</div>



Você trabalha na empresa de mineração OilyGiant. Sua tarefa é encontrar os melhores lugares para o desenvolvimento de novos poços de petróleo.

Para concluir essa tarefa, você vai precisar executar as seguintes etapas:

Ler os arquivos com parâmetros coletados de poços de petróleo na região selecionada: a quantidade de petróleo e o volume de reservas;
Construir um modelo para predizer o volume de reservas em novos poços;

Escolher os poços de petróleo que têm os maiores valores estimados;

Escolher a região com o maior lucro total para os poços de petróleo selecionados.

Você tem dados sobre amostras de petróleo de três regiões. Os parâmetros de cada poço de petróleo na região já são conhecidos. Construa um modelo que te ajudará a escolher a região com a margem de lucro mais alta. Utilize a técnica de Bootstrapping para analisar lucro potencial e riscos.

# abaixo está o enunciado de todo o projeto

2-Treine e teste o modelo para cada região em geo_data_0.csv:

Divida os dados em um conjunto de treinamento e um conjunto de validação em uma proporção de 75:25.

Treine o modelo e faça predições para o conjunto de validação.

Salve as predições e respostas corretas no conjunto de validação.

Imprima o volume médio previsto de reservas e o REQM do modelo.

Analise os resultados.

Coloque todos os passos anteriores em funções, execute os passos 2.1 a 2.5 nos arquivos 'geo_data_1.csv' e 'geo_data_2.csv'.

3-Prepare-se para o cálculo de lucro:

Armazene todos os valores necessários para os cálculos em variáveis separadas.

Dado o investimento de 100 milhões para 200 poços de petróleo, cada um precisa produzir, em média, uma quantidade de unidades equivalente a pelo menos 500 mil dólares para evitar prejuízos (isso é aproximadamente 111,1 unidades). Compare essa quantidade com o volume médio de cada região.

Forneça conclusões sobre a etapa de preparação para o cálculo de lucro.

4-Escreva uma função para calcular lucro de um conjunto de poços de petróleo selecionados e predições do modelo:

Escolha os 200 poços com os valores mais altos previstos de cada uma das 3 regiões (ou seja, arquivos 'csv').

Sumarize o volume alvo de reservas de acordo com essas predições. Armazene as predições para os 200 poços para cada uma das 3 regiões.

Calcule o lucro potencial dos 200 melhores poços por região. Apresente suas conclusões: sugira uma região para o desenvolvimento de poços de petróleo e justifique sua escolha.

5-Calcule riscos e lucro para cada região:

Usando as predições que você armazenou na etapa 4.2, use a técnica de bootstrapping com 1.000 amostras para encontrar a distribuição de lucros.

Encontre lucro médio, intervalo de confiança de 95% e o risco de prejuízo. Prejuízo é um lucro negativo, calcule-o como uma probabilidade e depois o expresse como uma porcentagem.

Apresente suas conclusões: sugira uma região para o desenvolvimento de poços de petróleo e justifique sua escolha. A sua escolha corresponde à escolha anterior na etapa 4.3?


<div class="alert alert-block alert-danger">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Vou responder por partes:
    
    - Você precisa ajustar um modelo usando regressão linear
    - Em seguida, você vai utilizar a técnica de bootstrapping que nos permite fazer inferência usando uma amostra. Veja que, dentre os 200 melhores poços, pode haver possos com prejuízo.
    - Deixei comentários mas nossa comunicação é limitada. Vocẽ programa bem. Vamos fazer uma iteração e caso haja dúvidas vou pedir um tutor para falar contigo.
</div>

<div class="alert alert-block alert-info">
<b>Eu achei o enunciado um pouco confuso, mas se eu entendi ele agora o bootstrap é para estar certo.</b> <a class="tocSkip"></a>
</div>

In [38]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression  
from sklearn.metrics import mean_squared_error
import numpy as np
from numpy.random import RandomState

In [39]:
df0 = pd.read_csv('geo_data_0.csv')
df1 = pd.read_csv('geo_data_1.csv')
df2 = pd.read_csv('geo_data_2.csv')


# Análise e limpeza de dados iniciais

# geo_data_0

In [40]:
df0.head()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

In [41]:
df0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [42]:
df0['product'].describe()

count    100000.000000
mean         92.500000
std          44.288691
min           0.000000
25%          56.497507
50%          91.849972
75%         128.564089
max         185.364347
Name: product, dtype: float64

In [43]:
df0.duplicated().sum()

0

**Todos os parametros dentro dos padrões para manipulação**

# geo_data_1

In [44]:
df1.head()

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

In [45]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [46]:
df1['product'].describe()

count    100000.000000
mean         68.825000
std          45.944423
min           0.000000
25%          26.953261
50%          57.085625
75%         107.813044
max         137.945408
Name: product, dtype: float64

In [47]:
df1.duplicated().sum()

0

**Todos os parametros dentro dos padrões para manipulação**

# geo_data_2

In [48]:
df2.head()

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

In [49]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [50]:
df2['product'].describe()

count    100000.000000
mean         95.000000
std          44.749921
min           0.000000
25%          59.450441
50%          94.925613
75%         130.595027
max         190.029838
Name: product, dtype: float64

In [51]:
df2.duplicated().sum()

0

**Todos os parametros dentro dos padrões para manipulação**

# Treino do modelo de regressão linear

In [52]:
features = df0.drop(['id','product'], axis = 1)
target = df0['product']

features_train, features_valid, target_train, target_valid = train_test_split(
    features, target, test_size=0.25, random_state=1
)
scaler = StandardScaler()
features_train = scaler.fit_transform(features_train) 
features_valid = scaler.transform(features_valid)

model = LinearRegression()  
model.fit(features_train , target_train)
x  = model.predict(features_valid)
xx = model.score(features_valid, target_valid)

media_volume = x.mean()
print('volume medio dos poços:' ,media_volume)

reqm = np.sqrt(mean_squared_error(target_valid , x))
print('erro médio da predição:',reqm)

volume medio dos poços: 92.49262459838863
erro médio da predição: 37.74258669996437


# Avaliação dos volumes médios dos poços 

In [53]:
def evaluate_region(df):
    features = df.drop(['id', 'product'], axis=1)
    target = df['product']
    
    features_train, features_valid, target_train, target_valid = train_test_split(
        features, target, test_size=0.25, random_state=1
    )
    
    scaler = StandardScaler()
    features_train_scaled = scaler.fit_transform(features_train)
    features_valid_scaled = scaler.transform(features_valid)
    
    model = LinearRegression()
    model.fit(features_train_scaled, target_train)
    predictions = model.predict(features_valid_scaled)
    
    mean_volume = predictions.mean()
    rmse = np.sqrt(mean_squared_error(target_valid, predictions))
    
    print(f"Volume médio dos poços: {mean_volume:.2f}")
    print(f"Erro médio da predição (REQM): {rmse:.2f}")
    
    
    return mean_volume, rmse





<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Não me parece haver problemas no teu código. 
</div>

<div class="alert alert-block alert-info">
<b>Ok, perdi bastante tempo tentando achar algum erro mas aparentemente não há mesmo.</b> <a class="tocSkip"></a>
</div>

In [54]:
evaluate_region(df1)

Volume médio dos poços: 69.12
Erro médio da predição (REQM): 0.89


(69.12040524285558, 0.8943375629130574)

In [55]:
df1['product'].describe()

count    100000.000000
mean         68.825000
std          45.944423
min           0.000000
25%          26.953261
50%          57.085625
75%         107.813044
max         137.945408
Name: product, dtype: float64

In [56]:
evaluate_region(df2)

Volume médio dos poços: 94.96
Erro médio da predição (REQM): 39.87


(94.9568304858529, 39.86671127773423)


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Os valores estão corretos/coerentes com outros projetos que já revisei</div>

**A região geo_data1 exibe as menores médias de volumes previstos, com um risco elevado de não alcançarmos a meta de 200 poços com produção mínima de 111 unidades (em milhares de barris). Esse cenário poderia resultar em retornos insuficientes, justificando a exclusão dessa região do plano de perfuração. Em contrapartida, as regiões geo_data0 e geo_data2 seriam selecionadas por apresentar uma média de volumes significativamente maior(as duas com valores bem próximos um do outro), oferecendo maior potencial de viabilidade econômica**

# Cálculo de lucro na perfuração de poços

In [57]:
def cal_profit(df):
    features = df.drop(['id', 'product'], axis=1)
    target = df['product']
    
    features_train, features_valid, target_train, target_valid = train_test_split(
        features, target, test_size=0.25, random_state=1)
    
    
    scaler = StandardScaler()
    features_train_scaled = scaler.fit_transform(features_train)
    features_valid_scaled = scaler.transform(features_valid)
    
    model = LinearRegression()
    model.fit(features_train_scaled, target_train)
    predictions = model.predict(features_valid_scaled)
    return pd.Series(predictions)
    
    
    



<div class="alert alert-block alert-danger">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

    - É importante dividir as tarefas: uma função para treinar e retornar as predições para todos os poços
    - Você pode criar uma função que apenas alcula lucro 
    - Você pode criar uma 3a função para calcular o boostrapping que recebe as predições (você fez abaixo)

        
</div>

<div class="alert alert-block alert-info">
<b>Deixei apenas as prediçoes na função e coloquei o calculo dos possíveis lucros nás respectivas células.</b> <a class="tocSkip"></a>
</div>

In [58]:
x0 = cal_profit(df0)

predict200 = x0.nlargest(200)
total200 = predict200.sum() 
profit_value = (total200 * 4500 / 1000000) - 100 


print(f"Somatório dos 200 principais poços previstos: {total200:.2f} mil barris")
print(f"Lucro potencial estimado: ${profit_value:.2f} milhoes")

Somatório dos 200 principais poços previstos: 31017.97 mil barris
Lucro potencial estimado: $39.58 milhoes


In [59]:
x1 = cal_profit(df1)

predict200 = x1.nlargest(200)
total200 = predict200.sum() 
profit_value = (total200 * 4500 / 1000000) - 100  


print(f"Somatório dos 200 principais poços previstos: {total200:.2f} mil barris")
print(f"Lucro potencial estimado: ${profit_value:.2f} milhoes")

Somatório dos 200 principais poços previstos: 27759.86 mil barris
Lucro potencial estimado: $24.92 milhoes


In [60]:
x2 = cal_profit(df2)

predict200 = x2.nlargest(200)
total200 = predict200.sum() 
profit_value = (total200 * 4500 / 1000000) - 100 


print(f"Somatório dos 200 principais poços previstos: {total200:.2f} mil barris")
print(f"Lucro potencial estimado: ${profit_value:.2f} milhoes")

Somatório dos 200 principais poços previstos: 29758.74 mil barris
Lucro potencial estimado: $33.91 milhoes


**A região 0 foi selecionada para este projeto por apresentar o maior potencial de rentabilidade. As previsões indicam que, ao escolher os 200 poços mais produtivos dessa área, há uma perspectiva significativamente maior de obtenção de lucros**

# Riscos e lucros para cada região com bootstraping

In [61]:
def bootstrap_profit(x):
    state = RandomState(12345)
    profits = []    
    for i in range(1000):
        
        sample = x.sample( replace=True, random_state=state.randint(0, 10000))
        
        sample_mean = sample.mean()
        profits.append(sample_mean)
   
    profits = pd.Series(profits)  
    ic_lower = profits.quantile(0.025)  
    ic_upper = profits.quantile(0.975)  
        
    profits_filtered = profits[(profits >= ic_lower) & (profits <= ic_upper)]
        
    print(f"Média das médias bootstrap: {profits_filtered.mean():.2f}")
    print(f"Intervalo de confiança 95%: [{ic_lower:.2f}, {ic_upper:.2f}]")
    
    return profits_filtered

**Nesta função, são geradas 1000 amostras, cada uma contendo 25000 poços, e calculada a média dos volumes de produção para cada amostra. O resultado é uma série de 1000 médias, que permite avaliar a viabilidade da perfuração na região com base na distribuição desses valores.**

<div class="alert alert-block alert-danger">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

    - Para o boostrapping, vocẽ precisa usar todos os poços não apenas os top 200. Talvez aqui esteja o ponto crítico. Veja que você está passando paenado apenas os top 200. Se você passar tudo e amostrar (ex: 200), haverá certamente poço com prejuízo.
    
</div>

<div class="alert alert-block alert-info">
<b>Agora fez mais sentido, mas pra mim o enunciado tinha ficado ambíguo. </b> <a class="tocSkip"></a>
</div>



<div class="alert alert-block alert-success">
<b> Comentário do revisor V2: </b> <a class="tocSkip"></a>

Correto. Deixo só uma observação quanto ao boostrapping. Segundo a literatura, o correto é usar todo o dataset em vez de subconjunto (n=200)
    
    
https://machinelearningmastery.com/a-gentle-introduction-to-the-bootstrap-method/

"The bootstrap sample is the same size as the original dataset. As a result, some samples will be represented multiple times in the bootstrap sample while others will not be selected at all."
</div>


In [62]:
x0 = cal_profit(df0)
x00 = bootstrap_profit(x0)
print(x00.sort_values())
lucros = (x00 * 4500 -  500000)
print('\n')
print('LUCRO/PREJUÍZO MÉDIO DAS AMOSTRAS')
lucros.sort_values()
print('\n')
maiores200 = lucros.nlargest(200)
maiores200.mean()

Média das médias bootstrap: 93.53
Intervalo de confiança 95%: [46.90, 140.71]
39      46.900879
966     47.074228
424     47.251301
381     47.425722
668     47.467020
          ...    
476    138.476018
167    138.584439
799    139.281448
196    140.713661
976    140.713661
Length: 951, dtype: float64


LUCRO/PREJUÍZO MÉDIO DE CADA AMOSTRA




55019.280179679765

In [63]:
x1 = cal_profit(df1)
x11 = bootstrap_profit(x1)
print(x11.sort_values())
lucros = (x11 * 4500 -  500000)
print('\n')
print('LUCRO/PREJUÍZO MÉDIO DAS AMOSTRAS')
lucros.sort_values()
print('\n')
maiores200 = lucros.nlargest(200)
maiores200.mean()

Média das médias bootstrap: 68.97
Intervalo de confiança 95%: [-0.04, 137.72]
763     -0.037945
175      0.066163
70       0.187790
48       0.218163
944      0.282706
          ...    
25     137.584677
909    137.604115
883    137.616301
665    137.697252
780    137.715572
Length: 950, dtype: float64


LUCRO/PREJUÍZO MÉDIO DE CADA AMOSTRA




71862.4016113067

In [64]:
x2 = cal_profit(df2)
x22 = bootstrap_profit(x2)
print(x22.sort_values())
lucros = (x22 * 4500 -  500000)
print('\n')
print('LUCRO/PREJUÍZO MÉDIO DAS AMOSTRAS')
lucros.sort_values()
print('\n')
maiores200 = lucros.nlargest(200)
maiores200.mean()

Média das médias bootstrap: 95.14
Intervalo de confiança 95%: [57.02, 134.40]
393     57.017633
448     57.082108
236     57.457773
715     57.457773
681     57.548493
          ...    
709    132.956517
662    133.560426
26     133.592681
368    134.301256
333    134.377721
Length: 950, dtype: float64


LUCRO/PREJUÍZO MÉDIO DE CADA AMOSTRA




38000.05202810771

**Realizamos um bootstrap com os dados previstos na etapa anterior, gerando médias a partir de 1000 amostras de 25000 poços cada. Pegamos as 200 maiores médias dessas 1000 amostras e achamos o lucro médio na perfuração de cada região.**

# Conclusão

**Ao aplicar a técnica de bootstrap na análise dos lucros previstos, obtivemos médias distintas das previamente analisadas, alterando a perspectiva inicial. Antes, a região 2 era a menos lucrativa, mas agora se destacou como a mais rentável com lucro médio de 71862.40(milhares) de dolares por 200 poços, sendo selecionada para o projeto de perfuração por oferecer o maior retorno financeiro.**